### ISSUE TITLE
Ordering of batches within a `vkQueueSubmit` call not specified.

### ISSUE THREAD

**Issue by NicolBolas** [KhronosGroup/Vulkan-Docs#371]
The specification for `VkSubmitInfo` makes it clear that the order that a batch's command buffers are submitted in is the order they are found in within the `pCommandBuffers` array.

However, the specification for `vkQueueSubmit` does not have a similar notation for `pSubmits`. That is, it is not clear whether each batch is executed in the order defined by `pSubmits`. If they're not executed in that order, then the specification should make that more explicit.

This is important for understanding issues like how pipeline barriers work between CBs in different batches. Issue #300 already explains that CBs in the same batch respect pipeline barriers, per the ordering of `pCommandBuffers`. And it respects the order that items are submitted to the same queue by separate `vkQueueSubmit` calls. But it would be odd if two batches in the same `vkQueueSubmit` call were unordered.

**Comment by TobiasHector** [comment]
I've got some edits which should clarify this to some degree incoming - but it does already say in the spec the following:

> Command buffer boundaries, both between primary command buffers of the same or different batches or submissions as well as between primary and secondary command buffers, do not introduce any implicit ordering constraints. In other words, submitting the set of command buffers (which can include executing secondary command buffers) between any semaphore or fence operations execute the recorded commands as if they had all been recorded into a single primary command buffer, except that the current state is reset on each boundary. Explicit ordering constraints can be expressed with events and pipeline barriers.

It's a bit wordy, and not particularly clear, but it's supposed to express that command buffers are submitted in order, as if they were a single command buffer, between any semaphores/fences (i.e. between submissions in a submit that don't have associated semaphores).

I have some edits incoming which hopefully clarify this with regards to pipeline barriers/events, but there's still more to do after that to really make it clear. At the moment I can't point to sections of the spec which spell out the answer to your question easily.

**Comment by oddhack** [comment]
This should be partly addressed in the 1.0.35 spec update, as part of the synchronization language rewrite.

**Comment by TobiasHector** [comment]
This should be fixed in the latest 1.0.41 specification, with everything now referring to [Submission Order](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#synchronization-submission-order). Feel free to re-open if you feel there's anything else that's still ambiguous.
